lib/checkother.cpp

```diff
@@ -1391,6 +1391,59 @@ void CheckOther::checkConstVariable()
         // Skip if address is taken
         if (Token::findmatch(var->nameToken(), "& %varid%", scope->bodyEnd, var->declarationId()))
             continue;
+        // Skip if another non-const variable is initialized with this variable
+        {
+            //Is it the right side of an initialization of a non-const reference
+            bool usedInAssignment = [&]
+                {
+                    auto findAssignment = [&](const Token* t) { return Token::findmatch(t, "%type% & %name% = %varid%", scope->bodyEnd, var->declarationId()); };
+                    for (const Token* match = findAssignment(var->nameToken()); match != nullptr; match = findAssignment(match->next()))
```

Comment:
I would personally prefer:

    bool usedInAssignment = false;
    for (const Token *tok = var->nameToken(); tok != scope->bodyEnd; tok = tok->next()) {
        if (!Token::Match(tok, "& %var% = %varid%", var->declarationId())
            continue;
        const Variable *refvar = tok->next()->variable();
        if (refvar && !refvar->isConst() && refvar->nameToken() == tok->next()) {
            usedInAssignment = true;
            break;
        }
    }